### PR TITLE
AI Chat: sniff subresource content via throttle to detect new content metadata for same-page navigations

### PR DESCRIPTION
**Pull request by petemill**
Resolves https://github.com/brave/brave-browser/issues/34945

- Creates throttle to intercept a specific subresource request
- Parses the subresource and shares transcript decision with existing JS global extraction method (which is also used for the initial page in the history)
- Creates a new mojom interface for the renderer to ~send the new content detected back to the browser~ notify the browser that page-changing content was detected. Actual content sending from renderer will only occur when asked for by browser during a user-initiated AIChat event.
- Tests for throttle and for transcript decision making

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
_Contains unit and browser tests._

- Navigate to youtube.com
- Navigate to a video (same-tab)
 - optional: ask Leo to summarize
- Navigate to another video (same-tab)
 - ask Leo to summarize
 Observe that with this PR, the correct video content is summarised and without then only the video content from the initially-navigated-to youtube page is available to Leo.

### CODE REVIEW

File: components/ai_chat/content/browser/page_content_fetcher.h
Comment (petemill) [inline components/ai_chat/content/browser/page_content_fetcher.h:27]: needs comment

File: components/ai_chat/content/browser/ai_chat_tab_helper.cc
Comment (petemill) [reply components/ai_chat/content/browser/ai_chat_tab_helper.cc:260]: Yes this likely needs changing - it's unclear to me but I think FetchPageContent could hold on to callback after TabHelper is destroyed.

File: components/ai_chat/content/browser/ai_chat_tab_helper.cc
Comment (petemill) [inline components/ai_chat/content/browser/ai_chat_tab_helper.cc:232]: Do not merge until this is resolved. Otherwise we'll be fetching the transcript content at the provided Urls on every relevant navigation, even when there's no active AI Chat conversation

Comment (petemill) [reply components/ai_chat/content/browser/ai_chat_tab_helper.cc]: This is resolved now - nothing will be parsed (or network fetched) until an AI Chat message is sent

File: components/ai_chat/content/browser/ai_chat_tab_helper.cc
Comment (petemill) [reply components/ai_chat/content/browser/ai_chat_tab_helper.cc]: this part is removed now

File: components/ai_chat/content/browser/page_content_fetcher.h
Comment (petemill) [reply components/ai_chat/content/browser/page_content_fetcher.h]: removed

**Comment by petemill** [comment]
Creating security review

File: components/ai_chat/renderer/ai_chat_resource_sniffer_url_loader.cc
Comment (petemill) [inline components/ai_chat/renderer/ai_chat_resource_sniffer_url_loader.cc:23]: Wasn't 100% sure on what to use for this part...

File: components/ai_chat/content/browser/ai_chat_tab_helper.cc
Comment (nullhook) [inline components/ai_chat/content/browser/ai_chat_tab_helper.cc:73]: i know you love `auto` but i can't infer what the type of `sender` is here.

File: components/ai_chat/content/browser/ai_chat_tab_helper.cc
Comment (nullhook) [inline components/ai_chat/content/browser/ai_chat_tab_helper.cc:78]: oh, it's `webcontents*`

File: components/ai_chat/renderer/page_content_extractor.cc
Comment (nullhook) [inline components/ai_chat/renderer/page_content_extractor.cc:89]: leftover?

File: components/ai_chat/renderer/page_content_extractor.cc
Comment (nullhook) [inline components/ai_chat/renderer/page_content_extractor.cc:116]: what is this type?

File: components/ai_chat/renderer/page_content_extractor.cc
Comment (nullhook) [inline components/ai_chat/renderer/page_content_extractor.cc:224]: why do you need to check if it's bound? `GetInterface` gurantees a bind, right? it's internally calling `BindNewEndpointAndPassReceiver` on a remote.

File: components/ai_chat/renderer/page_content_extractor.cc
Comment (petemill) [reply components/ai_chat/renderer/page_content_extractor.cc:116]: same as `intercepted_content_` which you also need to look where it's defined.

File: components/ai_chat/renderer/ai_chat_resource_sniffer_url_loader.h
Comment (iefremov) [inline components/ai_chat/renderer/ai_chat_resource_sniffer_url_loader.h:25]: there is an upcoming refactoring of the body sniffer, hopefully the new sniffer design is more logical than the current

File: components/ai_chat/renderer/yt_util.h
Comment (iefremov) [inline components/ai_chat/renderer/yt_util.h:32]: i think the parameter should be either const reference or string_view

File: components/ai_chat/renderer/yt_util.h
Comment (iefremov) [inline components/ai_chat/renderer/yt_util.h:28]: const ?

File: components/ai_chat/renderer/ai_chat_resource_sniffer_throttle_unittest.cc
Comment (boocmp) [inline components/ai_chat/renderer/ai_chat_resource_sniffer_throttle_unittest.cc:1]: I'm not entirely sure, but it looks like browser tests would be easier to write, also they can test some near to real api calls.

Comment (petemill) [reply components/ai_chat/renderer/ai_chat_resource_sniffer_throttle_unittest.cc:1]: I'm following upstreams example from `MimeSniffingThrottle`. And these tests seem successful in testing the parts we care about - whether the throttle was created and whether the delegate is called as expected. I think the team prefers unit tests to browser tests, in general - less flakey and more performant. Any issue with this?

File: components/ai_chat/renderer/ai_chat_resource_sniffer_throttle.cc
Comment (bridiver) [inline components/ai_chat/renderer/ai_chat_resource_sniffer_throttle.cc:32]: this kind of logging should normally be removed before merge per chromium guidelines

File: components/ai_chat/renderer/ai_chat_resource_sniffer_url_loader.h
Comment (iefremov) [reply components/ai_chat/renderer/ai_chat_resource_sniffer_url_loader.h:25]: sorry i've forgotten to link https://github.com/brave/brave-core/pull/21792/files

**Comment by petemill** [comment]
@iefremov @boocmp perhaps we can merge this before https://github.com/brave/brave-core/pull/21792 as this is a P2 issue and will need to be uplifted

File: components/ai_chat/content/browser/ai_chat_tab_helper.cc
Comment (petemill) [reply components/ai_chat/content/browser/ai_chat_tab_helper.cc:75]: Really? It's literally in rhs - `content::WebContents::From...`. That's why I've used `auto` here. It's unneccessary repetition in the same line.

**Review by bcaller — APPROVED** [review]
We looked over the PR together

File: components/ai_chat/renderer/page_content_extractor.cc
Comment (petemill) [reply components/ai_chat/renderer/page_content_extractor.cc:44]: We have to use the `ytplayer` JS object lookup for the initial page load, as the client page does not request the metadata from the API, and so there's nothing to intercept. For subsequent video loads, the opposite is true: 
the `ytplayer` JS object does not update, but the client page requests metadata from the API, and so we intercept it.

File: components/ai_chat/renderer/ai_chat_resource_sniffer_url_loader.cc
Comment (petemill) [reply components/ai_chat/renderer/ai_chat_resource_sniffer_url_loader.cc:72]: to be honest - I don't know if this is correct. @boocmp?

**Comment by petemill** [comment]
> We looked over the PR together

@bcaller who is "we"? 😄

File: components/ai_chat/renderer/page_content_extractor.h
Comment (iefremov) [inline components/ai_chat/renderer/page_content_extractor.h:74]: no need to initialize the unique pointer

File: components/ai_chat/core/browser/conversation_driver.h
Comment (iefremov) [inline components/ai_chat/core/browser/conversation_driver.h:148]: `techniquesused` typo

File: components/ai_chat/content/browser/ai_chat_tab_helper.cc
Comment (iefremov) [inline components/ai_chat/content/browser/ai_chat_tab_helper.cc:70]: maybe not for this PR, but chromium's general direction is to replace most DCHECKs with CHECKs, see https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/checks.md and the style guide

File: components/ai_chat/renderer/ai_chat_resource_sniffer_url_loader.cc
Comment (iefremov) [reply components/ai_chat/renderer/ai_chat_resource_sniffer_url_loader.cc:72]: the buffer is `std::string`

File: renderer/brave_url_loader_throttle_provider_impl.cc
Comment (iefremov) [inline renderer/brave_url_loader_throttle_provider_impl.cc:104]: why it is not just `SequencedTaskRunner`?

**Comment by iefremov** [comment]
in general the PR looks good to me, pls fix nits. Also the amount of logging probably could be decreased. It seems Pavel is ok to merge it before the body sniffer refactoring

File: renderer/brave_url_loader_throttle_provider_impl.cc
Comment (petemill) [reply renderer/brave_url_loader_throttle_provider_impl.cc:104]: That's what SpeedreaderThrottle had, but I suppose it has different requirements

File: components/ai_chat/renderer/page_content_extractor.cc
Comment (petemill) [reply components/ai_chat/renderer/page_content_extractor.cc:222]: removed the condition

File: renderer/brave_url_loader_throttle_provider_impl.cc
Comment (iefremov) [reply renderer/brave_url_loader_throttle_provider_impl.cc]: yes I think in the speedreader case we were worrying about running rust code on different threads

**Review by iefremov — APPROVED** [review]

**Comment by thypon** [comment]
good to go @petemill @diracdeltas

**Comment by kjozwiak** [comment]
Verification **PASSED** on `Win 11 x64` using the following build(s):

```

```

Using `1.66.24 Chromium: 123.0.6312.58` and the STR/Cases outlined via https://github.com/brave/brave-core/pull/22334#issue-2156189918, reproduced the issue where `Leo` wouldn't summarize YT videos correctly when you're summarizing several in a row. For example, in this case, it wouldn't summarize the video and started describing `Leo` (the feature) rather than using the videos transcripts to summarize what the user is currently watching:

reproducedLeo

**Pull request merged**
